File: packages/transport/src/file.ts
import { BaseTransport, type TransportQueueConfig } from './base.js'
import type { ARLSEvent } from '@agentlens/core'
import { promises as fs } from 'fs'
import { dirname, join } from 'path'

/**
 * Configuration for file transport
 */
export interface FileTransportConfig extends TransportQueueConfig {
  /** Path to write log file (required) */
  filePath: string
  /** Maximum file size in bytes before rotation (default: 50MB) */
  maxFileSize?: number
  /** Maximum number of rotated files to keep (default: 5) */
  maxRotatedFiles?: number
}

/**
 * File transport writes ARLS events to a JSONL file with rotation support.
 * Events are written one JSON object per line (JSONL format).
 * When file exceeds maxFileSize, it rotates: file.log → file.1.log, etc.
 *
 * @example
 * const transport = new FileTransport({
 *   filePath: './logs/agentlens.log',
 *   maxFileSize: 50 * 1024 * 1024, // 50MB
 *   maxRotatedFiles: 5,
 * })
 */
export class FileTransport extends BaseTransport {
  private filePath: string
  private maxFileSize: number
  private maxRotatedFiles: number
  private currentFileSize = 0

  constructor(config: FileTransportConfig) {
    super(config)
    this.filePath = config.filePath
    this.maxFileSize = config.maxFileSize ?? 50 * 1024 * 1024 // 50MB default
    this.maxRotatedFiles = config.maxRotatedFiles ?? 5
  }

  protected async drain(
    batch: Array<{ event: ARLSEvent; rendered: string }>
  ): Promise<void> {
    // Ensure directory exists
    const dir = dirname(this.filePath)
    try {
      await fs.mkdir(dir, { recursive: true })
    } catch {
      // Directory might already exist or creation might fail temporarily
    }

    for (const item of batch) {
      // Write one JSON line per event
      const line = `${item.rendered}\n`

      // Check if rotation is needed
      if (this.currentFileSize + line.length > this.maxFileSize) {
        await this.rotate()
        this.currentFileSize = 0
      }

      // Append to file
      try {
        await fs.appendFile(this.filePath, line, 'utf-8')
        this.currentFileSize += line.length
      } catch (error) {
        const message = error instanceof Error ? error.message : String(error)
        process.stderr.write(`[AgentLens] File write error: ${message}\n`)
      }
    }
  }

  /**
   * Rotate the log file when it exceeds maxFileSize.
   * Old files are renamed: agentlens.log → agentlens.1.log → agentlens.2.log, etc.
   */
  private async rotate(): Promise<void> {
    try {
      // Get the base name and extension
      const lastDot = this.filePath.lastIndexOf('.')
      const basePath = lastDot > 0 ? this.filePath.substring(0, lastDot) : this.filePath
      const ext = lastDot > 0 ? this.filePath.substring(lastDot) : ''

      // Rename existing rotated files in reverse order
      for (let i = this.maxRotatedFiles - 1; i >= 1; i--) {
        const oldPath = `${basePath}.${i}${ext}`
        const newPath = `${basePath}.${i + 1}${ext}`

        try {
          await fs.rename(oldPath, newPath)
        } catch {
          // File might not exist, that's okay
        }
      }

      // Rename current file to .1
      const newPath = `${basePath}.1${ext}`
      try {
        await fs.rename(this.filePath, newPath)
      } catch {
        // File might not exist yet
      }

      // Clean up oldest file if needed
      const oldestPath = `${basePath}.${this.maxRotatedFiles + 1}${ext}`
      try {
        await fs.unlink(oldestPath)
      } catch {
        // File doesn't exist, that's fine
      }
    } catch (error) {
      const message = error instanceof Error ? error.message : String(error)
      process.stderr.write(`[AgentLens] File rotation error: ${message}\n`)
    }
  }

  /**
   * Get the current file size to support testing.
   * @internal
   */
  async _getCurrentFileSize(): Promise<number> {
    try {
      const stats = await fs.stat(this.filePath)
      return stats.size
    } catch {
      return 0
    }
  }
}

File: python/agentlens-transport/agentlens_transport/file.py
"""
File transport for writing events to JSONL files with rotation.

Appends events to a file in JSONL format and rotates when size limit is reached.
"""

import sys
from pathlib import Path
from typing import Optional

import aiofiles

from agentlens_core import ARLSEvent

from agentlens_transport.base import BaseTransport, TransportConfig


class FileTransport(BaseTransport):
    """
    Transport that writes events to a JSONL file.

    Automatically rotates files when they exceed max_file_size.
    """

    def __init__(
        self,
        path: str = 'agentlens.log',
        max_file_size: int = 50 * 1024 * 1024,  # 50MB default
        max_rotations: int = 5,
        config: Optional[TransportConfig] = None,
    ) -> None:
        """
        Initialize file transport.

        Args:
            path: Path to log file (default: agentlens.log)
            max_file_size: Max file size before rotation (default: 50MB)
            max_rotations: Max number of rotated files to keep (default: 5)
            config: Transport configuration
        """
        super().__init__(config)
        self.path = Path(path)
        self.max_file_size = max_file_size
        self.max_rotations = max_rotations
        self.current_size = 0

        # Check existing file size
        if self.path.exists():
            self.current_size = self.path.stat().st_size

    async def drain(self, event: ARLSEvent, rendered: str) -> None:
        """
        Write event to file, rotating if necessary.

        Args:
            event: The ARLS event
            rendered: Pre-rendered JSON string (JSONL format)
        """
        try:
            # Check if we need to rotate
            await self._check_rotation()

            # Write event
            content = rendered + '\n'
            async with aiofiles.open(str(self.path), 'a') as f:
                await f.write(content)

            self.current_size += len(content.encode('utf-8'))

        except Exception as e:
            sys.stderr.write(f'[AgentLens] File write error: {e}\n')

    async def _check_rotation(self) -> None:
        """Check if file needs rotation and rotate if needed."""
        if self.path.exists():
            file_size = self.path.stat().st_size

            if file_size >= self.max_file_size:
                await self._rotate_files()

    async def _rotate_files(self) -> None:
        """Rotate log files."""
        try:
            # Remove old rotation if at max
            max_path = self.path.with_name(f'{self.path.stem}.{self.max_rotations}.log')
            if max_path.exists():
                max_path.unlink()

            # Shift existing rotations
            for i in range(self.max_rotations - 1, 0, -1):
                old_path = self.path.with_name(f'{self.path.stem}.{i}.log')
                new_path = self.path.with_name(f'{self.path.stem}.{i + 1}.log')

                if old_path.exists():
                    old_path.rename(new_path)

            # Rename current to .1
            if self.path.exists():
                self.path.rename(self.path.with_name(f'{self.path.stem}.1.log'))

            self.current_size = 0

        except Exception as e:
            sys.stderr.write(f'[AgentLens] File rotation error: {e}\n')

    async def close(self) -> None:
        """Close transport gracefully."""
        await super().close()


class CompactFileTransport(FileTransport):
    """
    File transport with smaller overhead.

    Useful for high-volume logging.
    """

    def __init__(
        self,
        path: str = 'agentlens.log',
        max_file_size: int = 10 * 1024 * 1024,  # 10MB default
        max_rotations: int = 3,
        config: Optional[TransportConfig] = None,
    ) -> None:
        """
        Initialize compact file transport.

        Args:
            path: Path to log file
            max_file_size: Max file size before rotation
            max_rotations: Max number of rotated files
            config: Transport configuration
        """
        super().__init__(path, max_file_size, max_rotations, config)
